harness/determined/metrics.py

```diff
@@ -0,0 +1,454 @@
+                network_throughput_collector.reset()
+                disk_collector.reset()
+
+
+
+
+            # Check if it is time to take a new measurement
+            if time.time() - last_measurement_time > self.MEASUREMENT_INTERVAL:
+                immutable_batch_idx = self.current_batch
+                cpu_util_measurement = cpu_util_collector.measure(immutable_batch_idx)
+                gpu_util_measurements = gpu_util_collector.measure(immutable_batch_idx)
+                net_thru_sent_measurement, net_thru_recv_measurement = network_throughput_collector.measure(immutable_batch_idx)
+                free_memory_measurement = free_memory_collector.measure(immutable_batch_idx)
+                disk_read_thru_measurement, disk_write_thru_measurement, iops_measurement = disk_collector.measure(immutable_batch_idx)
+
+                for gpu_uuid in gpu_util_measurements.keys():
+                    self.current_metrics.add_gpu_measurement(MetricsHolder.GPU_UTIL_METRIC, gpu_uuid, gpu_util_measurements[gpu_uuid])
+
+                self.current_metrics.add_nongpu_measurement(MetricsHolder.NET_THRU_SENT_METRIC, net_thru_sent_measurement)
+                self.current_metrics.add_nongpu_measurement(MetricsHolder.NET_THRU_RECV_METRIC, net_thru_recv_measurement)
+                self.current_metrics.add_nongpu_measurement(MetricsHolder.DISK_IOPS_METRIC, iops_measurement)
+                self.current_metrics.add_nongpu_measurement(MetricsHolder.DISK_THRU_READ_METRIC, disk_read_thru_measurement)
+                self.current_metrics.add_nongpu_measurement(MetricsHolder.DISK_THRU_WRITE_METRIC, disk_write_thru_measurement)
+                self.current_metrics.add_nongpu_measurement(MetricsHolder.FREE_MEM_METRIC, free_memory_measurement)
+                self.current_metrics.add_nongpu_measurement(MetricsHolder.SIMPLE_CPU_UTIL_METRIC, cpu_util_measurement)
+
+                last_measurement_time = time.time()
+
+
+            # Check if it is time to flush the batch and start a new batch
+            if time.time() - batch_start_time > self.FLUSH_INTERVAL:
+                self.enqueue_for_async_send(self.current_metrics.convert_to_post_format())
+                self.current_metrics.reset()
+                batch_start_time = time.time()
+
+            time.sleep(0.02)
+
+    def update_current_batch(self, new_current_batch):
+        self.current_batch = new_current_batch
+
+    def enqueue_for_async_send(self, metric_batch):
+        # This method can theoretically raise a FULL error, but SimpleQueues are unbounded so
+        # I don't think it will ever happen (https://docs.python.org/3/library/queue.html#queue.Queue.put)
+        # self.log("Enqueuing metric batch", metric_batch)
+        self.dispatch_queue.put_nowait(metric_batch)
+
+    def __enter__(self) -> "SystemMetricsThread":
+        self.start()
+        return self
+
+    def __exit__(self, *arg: Any) -> None:
+        self.quitting = True
```

Comment:
while you did show me that it's not required in python to join threads that you have started, can we join it anyway?  I feel like this breaks the basic principle of "cleanup after yourself".